diff --git a/dwm.c b/dwm.c
index 178ab7a..5822274 100644
--- a/dwm.c
+++ b/dwm.c
@@ -1866,6 +1866,9 @@ updatebars(void) {
 		.background_pixmap = ParentRelative,
 		.event_mask = ButtonPressMask|ExposureMask
 	};
+	XClassHint *ch = XAllocClassHint();
+	ch->res_name = "dwmstatus";
+	ch->res_class = "dwm";
 	for(m = mons; m; m = m->next) {
 		if (m->barwin)
 			continue;
@@ -1874,7 +1877,9 @@ updatebars(void) {
 		                          CWOverrideRedirect|CWBackPixmap|CWEventMask, &wa);
 		XDefineCursor(dpy, m->barwin, cursor[CurNormal]->cursor);
 		XMapRaised(dpy, m->barwin);
+		XSetClassHint(dpy, m->barwin, ch);
 	}
+	XFree(ch);
 }
 
 void
